src/relations/pgbouncer_provider.py

```diff
@@ -0,0 +1,265 @@
+
+        Generate password and handle user and database creation for the related application.
+        """
+        if not self._check_backend():
+            event.defer()
+            return
+        if not self.charm.unit.is_leader():
+            return
+
+        # Retrieve the database name and extra user roles using the charm library.
+        databases = event.database
+        extra_user_roles = event.extra_user_roles
+        rel_id = event.relation.id
+
+        # Creates the user and the database for this specific relation.
+        user = f"relation_id_{rel_id}"
+        password = pgb.generate_password()
+        try:
+            self.charm.backend.postgres.create_user(
+                user, password, extra_user_roles=extra_user_roles
+            )
+            for database in databases.split(","):
+                self.charm.backend.postgres.create_database(database, user)
+                # set up auth function
+                self.charm.backend.initialise_auth_function(dbname=database)
+        except (
+            PostgreSQLCreateDatabaseError,
+            PostgreSQLCreateUserError,
+            PostgreSQLGetPostgreSQLVersionError,
+        ) as e:
+            logger.exception(e)
+            self.charm.unit.status = BlockedStatus(
+                f"Failed to initialize {self.relation_name} relation"
+            )
+            return
+
+        self.charm.peers.add_user(user, password)
+
+        # Update pgbouncer config
+        cfg = self.charm.read_pgb_config()
+        cfg.add_user(user, admin=True if "SUPERUSER" in extra_user_roles else False)
+        self.update_postgres_endpoints(event.relation, cfg=cfg)
+        self.charm.render_pgb_config(cfg, reload_pgbouncer=True)
+
+        # Share the credentials and updated connection info with the client application.
+        self.database_provides.set_credentials(rel_id, user, password)
+        self.update_connection_info(event.relation)
+
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Remove the user created for this relation, and revoke connection permissions."""
+        if not self._check_backend():
+            event.defer()
+            return
+        if not self.charm.unit.is_leader():
+            return
+
+        cfg = self.charm.read_pgb_config()
+        database = self.get_database(event.relation)
+        cfg["databases"].pop(database, None)
+        cfg["databases"].pop(f"{database}_readonly", None)
+        user = f"relation_id_{event.relation.id}"
+        cfg.remove_user(user)
+        self.charm.render_pgb_config(cfg, reload_pgbouncer=True)
+
+        # Delete the user.
+        try:
+            self.charm.backend.postgres.delete_user(user)
+        except PostgreSQLDeleteUserError as e:
+            logger.exception(e)
+            self.charm.unit.status = BlockedStatus(
+                f"Failed to delete user during {self.relation_name} relation broken event"
+            )
+            raise
```

Comment:
Shouldn't you move the call about deleting the user from the database to before updating the config? So you can avoid raising the error again and just keep the blocked status (without changing anything in the pgbouncer config).